lib/Doctrine/ORM/AbstractQuery.php

```diff
@@ -417,11 +418,18 @@ public function processParameterValue($value)
             return $value;
         }
 
-        if (is_object($value) && $this->_em->getMetadataFactory()->hasMetadataFor(ClassUtils::getClass($value))) {
-            $value = $this->_em->getUnitOfWork()->getSingleIdentifierValue($value);
+        if (is_object($value)) {
+            try {
+                $value = $this->_em->getUnitOfWork()->getSingleIdentifierValue($value);
+
+                if ($value === null) {
+                    throw ORMInvalidArgumentException::invalidIdentifierBindingEntity();
+                }
+            } catch (MappingException $e) {
```

Comment:
Yes there is a performance impact because we are now trying to get the identifier value of every object processed by this method. And it tries to load the metadata everytime. However, doing this might be the only solution to actually check if an object is an entity or not. Unless someone has a better way of doing it, we have to choose between performance loss or not fixing this bug I guess.